aws/resource_aws_instance.go

```diff
@@ -1260,6 +1254,95 @@ func resourceAwsInstanceUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
+	if d.HasChange("root_block_device.0") && !d.IsNewResource() {
+		instance, err := resourceAwsInstanceFindByID(conn, d.Id())
+		if err != nil {
+			return fmt.Errorf("error retrieving instance %q: %w", d.Id(), err)
+		}
+		blockDevices, err := readBlockDevicesFromInstance(instance, conn)
+		if err != nil {
+			return fmt.Errorf("error retrieving volumes for instance %q: %w", d.Id(), err)
+		}
+		rd, ok := blockDevices["root"]
+		if !ok {
+			return fmt.Errorf("root volume no found for instance %q", d.Id())
+		}
+
+		rootDevice := rd.(map[string]interface{})
+		volumeID := rootDevice["volume_id"].(string)
```

Comment:
We should fetch the volume ID and other attributes from state (`d.Get()`) or diff (`d.GetChange()`), rather than using an additional API call 👍